Vegetarian-Food-Finder/Controller/HomeViewController.swift

```diff
@@ -19,39 +23,85 @@ final class HomeViewController: UIViewController, UITableViewDelegate, UITableVi
         super.viewDidLoad()
         tableView.dataSource = self
         tableView.delegate = self
+        tableView.rowHeight = UITableView.automaticDimension
+        getUserLocation()
         getResturantsData()
     }
 
+    private func getUserLocation() {
+        locationManager = CLLocationManager()
+        locationManager?.requestAlwaysAuthorization()
+        locationManager?.startUpdatingLocation()
+        locationManager?.delegate = self
+        locationManager?.allowsBackgroundLocationUpdates = true
+    }
+
     private func getResturantsData() {
         Task {
             do {
-                self.restaurants = try await YelpApi().searchVeggiBusinessesInSF()
+                self.restaurantsList = try await YelpApi().searchVeggiBusinessesInSF()
             } catch {
-                let alertController = UIAlertController(title: "Can get the data", message: "the server cannot process the request", preferredStyle: .alert)
-                let OKAction = UIAlertAction(title: "OK", style: .default, handler: nil)
-                alertController.addAction(OKAction)
-                self.present(alertController, animated: true, completion: nil)
+                showOkActionAlert(withTitle: "Can't get the data", andMessage: "the server cannot process the request")
             }
         }
     }
 
+    // Still working on this
+    private func recommendBasedOnPrice() -> [String: String] {
+        var restaurantsPrices: [String: String] = [:]
+        for restaurant in restaurantsList {
+            restaurantsPrices[restaurant.name] = restaurant.price
+        }
+        let sortedTubleArray = restaurantsPrices.sorted(by: { $0.value > $1.value })
+        let sortedRestaurantsPrices = sortedTubleArray.reduce(into: [:]) { $0[$1.0] = $1.1 }
+        return sortedRestaurantsPrices
+    }
+
+    private func recommendBasedOnRating() -> [String: Double] {
+        var restaurantsRatings: [String: Double] = [:]
+        for resturant in restaurantsList {
+            restaurantsRatings[resturant.name] = resturant.rating
+        }
+        let sortedTubleArray = restaurantsRatings.sorted(by: { $0.value > $1.value })
+        let sortedRestaurantsRatings = sortedTubleArray.reduce(into: [:]) { $0[$1.0] = $1.1 }
```

Comment:
Learn more in depth about higher order functions.
The concepts are not clear in why reduce is used here or how it works